dashboard/app/models/pd/workshop.rb

```diff
@@ -580,6 +580,34 @@ def attending_teachers
     sessions.flat_map(&:attendances).flat_map(&:teacher).uniq
   end
 
+  # Get all teachers who have attended all sessions of this workshop.
+  def teachers_attending_all_sessions(cdo_scholarship=false)
+    teachers_attending = sessions.flat_map(&:attendances).flat_map(&:teacher)
+
+    # Filter attendances to only scholarship teachers
+    if cdo_scholarship
+      teachers_attending.select! do |teacher|
+        Pd::ScholarshipInfo.exists?(
+          {
+            user: teacher,
+            application_year: school_year,
+            course: course_key
+          }
+        )
+      end
+    end
+
+    # Get number of sessions attended by teacher
+    attendance_count_by_teacher = Hash[
+      teachers_attending.uniq.map do |teacher|
+        [teacher, teachers_attending.count(teacher)]
+      end
+    ]
+
+    # Return only teachers who attended all sessions
+    attendance_count_by_teacher.select {|_, attendances| attendances == sessions.count}.keys
```

Comment:
In fact, this whole thing is quite the query optimization challenge!  Since this is just for payments I think you could merge this as-is, but it might be worth revisiting as a good exercise in query optimization within ActiveRecord.  Ask Will about his query-counting unit tests sometime.

Reply:
I know right, it's hairy! I'll do some research on optimizing queries here tomorrow and share with y'all to review.